ehrapy/api/_anndata_util.py

```diff
@@ -36,3 +36,66 @@ def get_column_values(adata: AnnData, indices: Union[int, List[int]]) -> np.ndar
         :class:`~numpy.ndarray` object containing the column values
     """
     return np.take(adata.X, indices, axis=1)
+
+
+def assert_encoded(adata: AnnData):
+    try:
+        assert "categoricals" in adata.uns_keys()
+    except AssertionError:
+        raise NotEncodedError from AssertionError("The AnnData object has not yet been encoded.")
+
+
+def get_numeric_vars(adata: AnnData) -> List[str]:
+    """Fetches the column names for numeric variables in X.
+
+    Args:
+        adata: :class:`~anndata.AnnData` object
+
+    Returns:
+        Set of column numeric column names
+    """
+
+    assert_encoded(adata)
+
+    return adata.uns["categoricals"]["not_categorical"]
+
+
+def set_numeric_vars(
+    adata: AnnData, values: np.ndarray, vars: Optional[List[str]] = None, copy: bool = False
+) -> Optional[AnnData]:
+    """Sets the column names for numeric variables in X.
+
+    Args:
+        adata: :class:`~anndata.AnnData` object
+        values: Matrix containing the replacement values
+        vars: List of names of the numeric variables to replace. If `None` they will be detected using ~ehrapy.pp.get_numeric_vars.
+        copy: Whether to return a copy with the normalized data.
+
+    Returns:
+        :class:`~anndata.AnnData` object with updated X
+    """
+
+    if vars is None:
+        vars = get_numeric_vars(adata)
+
+    if not np.issubdtype(values.dtype, np.number):
+        raise TypeError(f"values must be numeric (current dtype is {values.dtype})")
+
+    n_values = values.shape[1]
+
+    if n_values != len(vars):
+        raise ValueError(f"Number of values ({n_values}) does not much number of vars ({len(vars)})")
+
+    if copy:
+        adata = adata.copy()
+
+    vars_idx = get_column_indices(adata, vars)
+
+    for i in range(n_values):
+        adata.X[:, vars_idx[i]] = values[:, i]
+
+    return adata
```

Comment:
I am not sure whether you want to return the whole AnnData object here.

What I would have expected is something like a dictionary that looks like 

```
{ 
    numerical_feature_1_var_name: numpy column with values,
    numerical_feature_2_var_name: numpy column with values,
    ...
}
```

And then later you replace the respective columns in the actual AnnData object when normalizing. My idea might however be slow and space wasteful. Not sure what your plan is.

This is just one of many ideas to approach this. Feel completely free to follow your own ideas here!

Reply:
I was going to calculate the normalised values and then use this function to set them. I'll probably implement it that way and then we can discuss if something else would be better.

Reply:
`I was going to calculate the normalised values and then use this function to set them. I'll probably implement it that way and then we can discuss if something else would be better.`

Just need to ensure that you still have all previous non-numerical values and that you don't change the order of the columns. But if it works out totally fine.

Reply:
Ah, this is something I haven't thought of before. Should the normalized values be added as new columns of `X` or should they replace the existing columns and store the old values somewhere (I was going to do this)?

Reply:
Maybe we can store the unnormalized data in a new layer -> layer "raw"?
Then you can either replace the columns one by one in the current X with the normalized values or you replace the complete X. In both cases we need to ensure that the order still matches.

`X` should always have the same dimension. Normalized and unnormalized values should never be in the same matrix.

Reply:
Can't we just use our existing layer "original"? It should contain all original numerical values (and those, that are encoded are already stored in _.uns_).

Reply:
Guess this could also work. Wasn't quite sure whether this is issue because of quality control etc. But QC should affect the layer as well right?

Reply:
Per default yes, IMO. (But we can also allow passing a parameter like `copy_layer=True` or similar, allowing API users decide on whether changes made on `X` should be reflected in the layer or not)

Reply:
I think that it would be easiest for @lazappi if we he simply adds a `raw` layer as discussed and we, @Imipenem will later discuss how we can optimize the required memory of the layers.

But I'd [KISS ](https://en.wikipedia.org/wiki/KISS_principle) it for now.